src/axolotl/utils/models.py

```diff
@@ -324,671 +324,826 @@ def load_processor(cfg: DictDefault, tokenizer: PreTrainedTokenizerBase):
+        if self.cfg.adapter in ["lora", "qlora"]:
+            needs_fa2_dtype = True
+            if self.cfg.gradient_checkpointing:
+                self.model.gradient_checkpointing_enable(
+                    gradient_checkpointing_kwargs=self.cfg.gradient_checkpointing_kwargs
+                )
+
+        self.prepare_model(qlora_fsdp)
 
-    if torch.cuda.device_count() > 1 and int(os.getenv("WORLD_SIZE", "1")) == 1:
-        setattr(model, "is_parallelizable", True)
-        setattr(model, "model_parallel", True)
+        # LlamaRMSNorm layers are in fp32 after kbit_training or full finetune, so we need to
+        # convert them back to fp16/bf16 for flash-attn compatibility.
+        if (needs_fa2_dtype or self.cfg.flash_attention) and not qlora_fsdp:
+            LOG.info(
+                "converting modules to %s for flash attention", self.cfg.torch_dtype
+            )
+            self.convert_embedding_modules_dtype(
+                embedding_modules,
+                dist_dtype=self.cfg.torch_dtype,
+                before_kbit_train_or_finetune=False,
+            )
+
+        # ---------------------------------------------------------
+        #  load lora or adapter
+        # ---------------------------------------------------------
+        lora_config = None
+        if not self.reference_model or self.cfg.lora_model_dir:
+            # if we're not loading the reference model, then we're loading the model for training
+            # then the dpo trainer doesn't want the peft model loaded over it, it just wants the lora/peft config
+            if (
+                self.cfg.adapter
+                and self.cfg.rl in ["dpo", "ipo", "kto"]
+                and not self.cfg.merge_lora
+            ):
+                _, lora_config = load_lora(
+                    self.model, self.cfg, inference=False, config_only=True
+                )
+            else:
+                self.model, lora_config = load_adapter(
+                    self.model, self.cfg, self.cfg.adapter
+                )
 
-    requires_grad = []
-    for name, param in model.named_parameters(recurse=True):
-        if param.requires_grad:
-            requires_grad.append(f"{name}: {param.requires_grad}")
-    if len(requires_grad) == 0:
-        LOG.warning("there are no parameters that require gradient updates")
-    if hasattr(model, "config"):
-        model.config.use_cache = False
+        # ---------------------------------------------------------
+        #  put model to accelerator
+        # ---------------------------------------------------------
+        if (
+            self.cfg.ddp
+            and not (
+                "load_in_8bit" in self.model_kwargs
+                and self.model_kwargs["load_in_8bit"]
+            )
+            and not (
+                self.cfg.rl
+                and "load_in_4bit" in self.model_kwargs
+                and self.model_kwargs["load_in_4bit"]
+            )
```

Comment:
Minor difference here. This is not equivalent to 
```
and not load_in_8bit  # This is cfg.load_in_8bit
and not (cfg.rl and cfg.load_in_4bit)  # This is cfg.load_in_4bit
```

as current check uses the below kwarg for comparison 
```
self.model_kwargs["load_in_8bit"] = self.cfg.load_in_8bit and self.cfg.adapter is not None
```

However, there should be no issue as a user should always have an adapter when `load_in*` is used.

Maybe this part can stripped down a bit to use the below like in `prepare_model` for readability
```
        is_load_in_8bit = (
            "load_in_8bit" in self.model_kwargs and self.model_kwargs["load_in_8bit"]
        )
        is_load_in_4bit = (
            "load_in_4bit" in self.model_kwargs and self.model_kwargs["load_in_4bit"]
        )
```